src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -272,14 +279,32 @@ private void OnRead(UvStreamHandle handle, int status)
                 }
 
                 error = new IOException(uvError.Message, uvError);
+                _currentWritableBuffer?.Commit();
             }
+            else
+            {
+                Debug.Assert(_currentWritableBuffer != null);
 
-            Input.IncomingComplete(readCount, error);
+                var currentWritableBuffer = _currentWritableBuffer.Value;
+                currentWritableBuffer.Advance(readCount);
+                var flushTask = currentWritableBuffer.FlushAsync();
+                if (!flushTask.IsCompleted)
+                {
+                    ((IConnectionControl) this).Pause();
+                    if (await flushTask)
+                    {
+                        Thread.Post(connectionControl => connectionControl.Resume(), (IConnectionControl)this);
```

Comment:
Once https://github.com/aspnet/KestrelHttpServer/pull/1338 is merged you can remove the Thread.Post here.